File: LinkedList/21. Merge Two Sorted Lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function (list1, list2) {
    if (list1 == null && list2 == null) {
        return list1;
    }

    if (list1 == null && list2 != null) {
        return list2;
    }

    var headNewList = new ListNode(0);
    var newList = headNewList;

    while (list1 != null && newList != null && list2 != null) {

        if (parseInt(list1.val) < parseInt(list2.val)) {
            newList.next = list1;
            list1 = list1.next;
        } else {
            newList.next = list2;
            list2 = list2.next;
        }
        newList = newList.next;
    }

    if (list2 == null) {
        newList.next = list1;
    }

    if (list1 == null) {
        newList.next = list2;
    }

    return headNewList.next;
};

File: NeetCode 150/21. Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        
        dummy_node = ListNode(0)
        current_node = dummy_node
        while list1 or list2:
            if list1 and list2:
                if list1.val > list2.val:
                    current_node.next = list2
                    current_node = current_node.next
                    list2 = list2.next
                else:
                    current_node.next = list1
                    current_node = current_node.next
                    list1 = list1.next

            if not list1 and list2:
                while list2:
                    current_node.next = list2
                    current_node = current_node.next

                    list2 = list2.next

            if list1 and not list2:
                while list1:
                    current_node.next = list1
                    current_node = current_node.next
                    list1 = list1.next

            
        return dummy_node.next

